### ISSUE TITLE
Linker errors for hwy::SupportedTargets(void) and hwy::GetChosenTarget(void)

### ISSUE THREAD

**Issue by OliverWangData** [google/highway#2378]
Hey all,

I'm building highway as a static library to link into UE. I'm using the following configuration:

```
[path to cmake]
-G "Visual Studio 17 2022" 
-DCMAKE_BUILD_TYPE=Release 
-DBUILD_SHARED_LIBS=OFF 
-DCMAKE_INSTALL_PREFIX=E:\Unreal\[name] 
-DCMAKE_INSTALL_INCLUDEDIR=E:\Unreal\[name]\Source\ThirdParty\Highway\Include 
-DCMAKE_INSTALL_BINDIR=E:\Unreal\[name]\Binaries\ThirdParty\Highway 
-DCMAKE_INSTALL_LIBDIR=E:\Unreal\[name]\Libraries\ThirdParty\Highway 
-B E:\Unreal\[name]\Intermediate\ThirdParty\Highway 
E:\Unreal\[name]\Source\ThirdParty\Highway\highway-1.2.0

[path to cmake] --build E:\Unreal\[name]\Intermediate\ThirdParty\Highway --config Release -j
[path to cmake] --install E:\Unreal\[name]\Intermediate\ThirdParty\Highway --config Release
```

This is on Windows 11, with the 3 installed .libs properly linked in the project.

Currently, I'm getting the following errors indicating that the two functions in the title are unresolved. This happens on numerous configurations that I've tried. 

```
2>[script].cpp.obj : error LNK2019: unresolved external symbol "__int64 __cdecl hwy::SupportedTargets(void)" (?SupportedTargets@hwy@@YA_JXZ) referenced in function "public: static void __cdecl hwy::FunctionCache<void>::ChooseAndCall<&void (__cdecl*const * const SIMD::TestFuncHighwayDispatchTable)(void)>(void)" (??$ChooseAndCall@$1?TestFuncHighwayDispatchTable@SIMD@@3QBQ6AXXZB@?$FunctionCache@X$$V@hwy@@SAXXZ)
2>[script].cpp.obj : error LNK2019: unresolved external symbol "struct hwy::ChosenTarget & __cdecl hwy::GetChosenTarget(void)" (?GetChosenTarget@hwy@@YAAEAUChosenTarget@1@XZ) referenced in function "public: static void __cdecl hwy::FunctionCache<void>::ChooseAndCall<&void (__cdecl*const * const SIMD::TestFuncHighwayDispatchTable)(void)>(void)" (??$ChooseAndCall@$1?TestFuncHighwayDispatchTable@SIMD@@3QBQ6AXXZB@?$FunctionCache@X$$V@hwy@@SAXXZ)
```

Dumping hwy.lib  with dumpbin /SYMBOLS reveals the following:

```
...
    Section length  213, #relocs    6, #linenums    0, checksum C943BB6A, selection    1 (pick no duplicates)
015 00000000 UNDEF  notype       External     | __imp___stdio_common_vfprintf
016 00000000 SECT9  notype ()    External     | ?SupportedTargets@hwy@@YA_JXZ (__int64 __cdecl hwy::SupportedTargets(void))
...

01F 00000000 SECT9  notype       Label        | $LN91
020 00000000 SECTA  notype       Static       | .xdata
    Section length   10, #relocs    1, #linenums    0, checksum 7CF4C652, selection    5 (pick associative Section 0x9)
022 00000000 SECTA  notype       Static       | $unwind$?SupportedTargets@hwy@@YA_JXZ
023 00000000 SECTB  notype       Static       | .pdata
    Section length    C, #relocs    3, #linenums    0, checksum 205670FA, selection    5 (pick associative Section 0x9)
025 00000000 SECTB  notype       Static       | $pdata$?SupportedTargets@hwy@@YA_JXZ
026 00000000 SECTC  notype       Static       | .xdata
    Section length   14, #relocs    3, #linenums    0, checksum E1ACCE8D, selection    5 (pick associative Section 0x9)
028 00000000 SECTC  notype       Static       | $chain$0$?SupportedTargets@hwy@@YA_JXZ
029 00000000 SECTD  notype       Static       | .pdata
    Section length    C, #relocs    3, #linenums    0, checksum D4077F8D, selection    5 (pick associative Section 0x9)
02B 00000000 SECTD  notype       Static       | $pdata$0$?SupportedTargets@hwy@@YA_JXZ
02C 00000000 SECTE  notype       Static       | .xdata
    Section length   10, #relocs    3, #linenums    0, checksum 99019A16, selection    5 (pick associative Section 0x9)
02E 00000000 SECTE  notype       Static       | $chain$1$?SupportedTargets@hwy@@YA_JXZ
02F 00000000 SECTF  notype       Static       | .pdata
    Section length    C, #relocs    3, #linenums    0, checksum 721FFA27, selection    5 (pick associative Section 0x9)
031 00000000 SECTF  notype       Static       | $pdata$1$?SupportedTargets@hwy@@YA_JXZ
032 00000000 SECT10 notype       Static       | .voltbl
    Section length    2, #relocs    0, #linenums    0, checksum 7F2364FC, selection    5 (pick associative Section 0x9)
034 00000000 SECT10 notype       Static       | _volmd
035 00000000 SECT11 notype       Static       | .voltbl
...
056 00000000 SECT25 notype ()    External     | ?HaveFloat16@hwy@@YA_NXZ (bool __cdecl hwy::HaveFloat16(void))
057 00000000 SECT26 notype ()    External     | ?HaveFloat64@hwy@@YA_NXZ (bool __cdecl hwy::HaveFloat64(void))
058 00000000 UNDEF  notype ()    External     | ?SupportedTargets@hwy@@YA_JXZ (__int64 __cdecl hwy::SupportedTargets(void))
059 00000000 UNDEF  notype ()    External     | ?GetChosenTarget@hwy@@YAAEAUChosenTarget@1@XZ (struct hwy::ChosenTarget & __cdecl hwy::GetChosenTarget(void))
05A 00000000 SECT1F notype ()    External     | ?GetTarget@N_SSSE3@hwy@@YA_JXZ (__int64 __cdecl hwy::N_SSSE3::GetTarget(void))
05B 00000000 SECT24 notype ()    External     | ?GetVectorBytes@N_SSSE3@hwy@@YA_KXZ (unsigned __int64 __cdecl hwy::N_SSSE3::GetVectorBytes(void))
...

```

In particular, the "UNDEF" seems to imply something may incorrect with my build setup. I checked the intermediate targets.obj as well, but it appears that the .obj has those functions correctly defined.

```
...
013 00000000 SECT9  notype       Static       | .text$mn
    Section length  213, #relocs    6, #linenums    0, checksum C943BB6A, selection    1 (pick no duplicates)
015 00000000 UNDEF  notype       External     | __imp___stdio_common_vfprintf
016 00000000 SECT9  notype ()    External     | ?SupportedTargets@hwy@@YA_JXZ (__int64 __cdecl hwy::SupportedTargets(void))
017 00000000 SECT5  notype ()    External     | ?DisableTargets@hwy@@YAX_J@Z (void __cdecl hwy::DisableTargets(__int64))
018 00000000 SECT8  notype ()    External     | ?SetSupportedTargetsForTest@hwy@@YAX_J@Z (void __cdecl hwy::SetSupportedTargetsForTest(__int64))
019 00000000 SECT7  notype ()    External     | ?GetChosenTarget@hwy@@YAAEAUChosenTarget@1@XZ (struct hwy::ChosenTarget & __cdecl hwy::GetChosenTarget(void))
01A 00000000 SECT6  notype ()    Static       | ?FlagsFromCPUID@x86@?A0xd176fdb1@hwy@@YA_KXZ (unsigned __int64 __cdecl hwy::`anonymous namespace'::x86::FlagsFromCPUID(void))
...
```

I'm a bit at a loss at what to do. Any ideas?

**Comment by johnplatts** [comment]
> I'm a bit at a loss at what to do. Any ideas?

Need to link against hwy.lib (which contains the `hwy::SupportedTargets` and `hwy::GetChosenTarget` functions) to fix the linker error.

**Comment by OliverWangData** [comment]
I actually am doing that right now. The libraries are installed to E:\Unreal\[name]\Libraries\ThirdParty\Highway, and I've got this in my build.cs (For linking to):

```
Type = ModuleType.External;

string libraryName = "Highway";
    string projectDirectory = Path.GetDirectoryName(Target.ProjectFile.FullName);
    string libraryDirectory = Path.Combine(
        projectDirectory, "Libraries", "ThirdParty", libraryName
        );

    // See pkgconfig in the libraries path
    PublicDefinitions.Add("HWY_STATIC_DEFINE");
    PublicDefinitions.Add("HWY_TEST_STANDALONE=1");

    PublicSystemIncludePaths.Add(Path.Combine(ModuleDirectory, "Include"));
PublicAdditionalLibraries.Add(Path.Combine(libraryDirectory, "hwy.lib"));
PublicAdditionalLibraries.Add(Path.Combine(libraryDirectory, "hwy_contrib.lib"));
PublicAdditionalLibraries.Add(Path.Combine(libraryDirectory, "hwy_test.lib"));
```

In particular, the additional libraries paths expand out to E:\Unreal\[name]\Libraries\ThirdParty\Highway\[libname].lib. The public definitions come from the pkconfig found in the install output, but with them added or not the error still persists.

**Comment by OliverWangData** [comment]
Hey johnplatts, been busy for the past few days but thanks for the reply!

So I set up your test program and ran it, and it did manage to link those successfully. So I tested in my UE environment, those two functions did not, but I also searched for hwy functions which did not have any "UNDEF" symbols and they also did not link properly. 

Long story short, after testing various things including messing around with compiler flags and learning a ton about the build process trying to debug it, it turns out there's a subtle-ish bug with unreal module dependencies. If a module is a public dependency for another, it should propagate over when that other module is also a dependency. But it wasn't, so the highway third party module wasn't propagating over, leading to the unresolved symbols. 

I think I was confused because the two functions which I ran into were the undef ones which convinced me there must've been something wrong with the .lib and I didn't know enough about the c++ compiling process to question it. 

Regardless, thanks for your time.
